### ISSUE TITLE
Conditional mapped type changes depending on last saved file

### ISSUE THREAD

**Issue by MartinJohns** [microsoft/TypeScript#40868]
I encountered a weird issue that I can't make much sense of. Depending on which file I saved last the `MergeProperty<>` type has a different signature. Maybe I'm doing something really wrong or there's an easier way to achieve this behavior.

**TypeScript Version:**  4.1.0-dev.20201001

**Search Terms:** `keyof index`

**Code**
Repository to repreduce the issue: https://github.com/MartinJohns/typescript-bug-signature

```ts
// utils.ts

export type First<T extends any[]> = T[0];
export type DropFirst<T extends any[]> = T extends readonly [any?, ...infer U] ? U : [...T];

export type GetProperty<T, N extends string> = T extends { [P in N]: infer R } ? R : never;
export type SetProperty<T, N extends string, V> = { readonly [P in keyof T | N]: P extends N ? V : P extends keyof T ? T[P] : never; };
export type MergeProperty<T, N extends string, V> =
  N extends keyof T ?
    SetProperty<T, N, {
      readonly [P in (keyof GetProperty<T, N>) | (keyof V)]: P extends keyof V ? V[P] : P extends keyof GetProperty<T, N> ? GetProperty<T, N>[P] : never
    }>
  : SetProperty<T, N, { readonly [P in keyof V]: V[P] }>;


// index.ts

import { DropFirst, First, MergeProperty } from './utils';

export type AddCreatedProperty<TName extends string, TProperty> = { type:'addCreatedProperty'; name: TName; creator: () => TProperty }
export type AddStaticProperty<TName extends string, TProperty> = { type: 'addStaticProperty', name: TName; value: TProperty }
export type UseResult<TResult> = { type: 'useResult'; handler: (result: TResult) => void }

export type Step =
    | AddCreatedProperty<any, any>
    | AddStaticProperty<any, any>
    | UseResult<any>

type Result<T extends Step[], R> = 
    First<T> extends undefined ? R :
    First<T> extends AddCreatedProperty<infer TName, infer TType> ? Result<DropFirst<T>, MergeProperty<R, 'data', { [P in TName]: TType }>> :
    First<T> extends AddStaticProperty<infer TName, infer TType> ? Result<DropFirst<T>, MergeProperty<R, 'data', { [P in TName]: TType }>> :
    Result<DropFirst<T>, R>

class Builder<T extends Step[]> {
    constructor(private readonly steps: T) {}

    append<S extends Step>(step: S): Builder<[...T, S]> { return new Builder([...this.steps, step]); }

    addCreatedProperty = <TName extends string, TProp>(name: TName, creator: () =>
        this.append({ type: 'addCreatedProperty', name, creator });

    addStaticProperty = <TName extends string, TProp>(name: TName, value: TProp) =>
        this.append({ type: 'addStaticProperty', name, value });

    result = (handler: (result: Result<T, {}>) => void) =>
        this.append({ type: 'useResult', handler });
}

new Builder<[]>([])
    .addCreatedProperty('first', () => 1)
    .addCreatedProperty('another', () => 2)
    .result(result => result.data.first)
```

**Expected behavior:**

`MergeProperty<>` contains `{ readonly [P in keyof GetProperty<T, N> | keyof V]: P extends ... }` regardless which file was saved last.

**Actual behavior:**

`MergeProperty<>` contains `{ readonly [x: string]: any extends ... }` when `index.ts` was last saved, and `{ readonly [P in keyof GetProperty<T, N> | keyof V]: P extends ... }` when `utils.ts` was last saved.

![example](https://user-images.githubusercontent.com/5269069/94789203-e97a7980-03d4-11eb-9ea2-5d3768b770de.gif)

**Playground Link:** Not applicable, as this behavior requires multiple files.

**Related Issues:**

- #40206 perhaps

I tried to simplify the example as much as possible (and I'm still trying to simplify it further).

**Comment by MartinJohns** [comment]
The generated `.d.ts` file contains the expected type.

**Comment by andrewbranch** [comment]
I haven't been able to repro this from your example 🤔 Can you add exact steps on how to reproduce? It’s a little hard to tell with the looping gif, but I think I'm doing what you're doing.

**Comment by MartinJohns** [comment]
The loop is a bit fast, but basically all I did was:

- Modify and save `utils.ts`.
- Hover over `MergeProperty<>` -> has the correct type.
- Open `index.ts`.
- Hover over `result.data` -> it has the right properties.
- Modify and save `index.ts`.
- Hover over `result.data` again -> it now has the wrong type.

---

But strangely... I can't reproduce it anymore. 🤔

Three days ago VS Code 1.49.3 was released and it got updated automatically within my WSL2 instance. Perhaps this updated fixed the issue, tho I couldn't explain how - especially given the use of the workspace TypeScript version, not the VS Code one. I'm not certain anymore which version of VS Code I had exactly when I created this issue.

I'll close the issue as the problem vanished.
